### PR TITLE
Add NLI models

### PR DESCRIPTION
**Pull request by ZhaofengWu**
SNLI: dev 92.47, test 91.65, URL https://storage.googleapis.com/allennlp-public-models/snli-roberta-large-2020.02.27.tar.gz

MNLI: dev-matched 89.10, dev-mismatched 88.71, URL https://storage.googleapis.com/allennlp-public-models/mnli-roberta-large-2020.02.27.tar.gz

### CODE REVIEW

File: allennlp/data/dataset_readers/snli.py [hunk 1]
```diff
@@ -68,16 +68,23 @@ def text_to_instance(
     ) -> Instance:
 
         fields: Dict[str, Field] = {}
-        premise_tokens = self._tokenizer.tokenize(premise)
-        hypothesis_tokens = self._tokenizer.tokenize(hypothesis)
-        fields["premise"] = TextField(premise_tokens, self._token_indexers)
-        fields["hypothesis"] = TextField(hypothesis_tokens, self._token_indexers)
+
+        if isinstance(self._tokenizer, PretrainedTransformerTokenizer):
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/snli.py]: Slight preference for having this be a configuration option instead of relying on this type check.

File: allennlp/data/dataset_readers/snli.py [hunk 2]
```diff
@@ -29,14 +29,29 @@ class SnliReader(DatasetReader):
         We use this `Tokenizer` for both the premise and the hypothesis.  See :class:`Tokenizer`.
     token_indexers : `Dict[str, TokenIndexer]`, optional (default=`{"tokens": SingleIdTokenIndexer()}`)
         We similarly use this for both the premise and the hypothesis.  See :class:`TokenIndexer`.
+    tokenize_separately : `bool`, optional
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/snli.py]: I like having the default be a best guess like this, though the name is a bit confusing to me.  Maybe `combine_fields` or `use_single_field`, or something, with the default being `False`?

Comment (matt-gardner) [reply allennlp/data/dataset_readers/snli.py]: The issue is this isn't necessarily about tokenization, it's more about how the data is represented.

Comment (ZhaofengWu) [reply allennlp/data/dataset_readers/snli.py]: Do you mean the default being `None`? Because if it's `False` we can no longer take a guess.

Comment (matt-gardner) [reply allennlp/data/dataset_readers/snli.py]: Right, sorry, I guess I just meant negating the logic in the guess.

File: allennlp/data/dataset_readers/snli.py [hunk 3]
```diff
@@ -29,14 +29,27 @@ class SnliReader(DatasetReader):
         We use this `Tokenizer` for both the premise and the hypothesis.  See :class:`Tokenizer`.
     token_indexers : `Dict[str, TokenIndexer]`, optional (default=`{"tokens": SingleIdTokenIndexer()}`)
         We similarly use this for both the premise and the hypothesis.  See :class:`TokenIndexer`.
+    combine_input_fields : `bool`, optional
+            (default=`not isinstance(tokenizer, PretrainedTransformerTokenizer)`)
```
Comment (matt-gardner) [inline allennlp/data/dataset_readers/snli.py]: This line is now wrong.  
```suggestion
            (default=`isinstance(tokenizer, PretrainedTransformerTokenizer)`)
```
After this, looks great!

File: training_config/mnli_roberta.jsonnet [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    },
+    "feedforward": {
+      "input_dim": transformer_dim,
+      "num_layers": 1,
+      "hidden_dims": transformer_dim,
+      "activations": "tanh"
+    },
+    "dropout": 0.1
+  },
+  "iterator": {
```
Comment (matt-gardner) [inline training_config/mnli_roberta.jsonnet]: Sorry I didn't notice this earlier - this needs to be updated to match #3700.  We can probably also just remove the `sorting_keys` key here.

File: training_config/mnli_roberta.jsonnet [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+          "max_length": 512
+        }
+      }
+    },
+    "seq2vec_encoder": {
+       "type": "cls_pooler",
+       "embedding_dim": transformer_dim,
+       "cls_is_last_token": cls_is_last_token
+    },
+    "feedforward": {
```
Comment (matt-gardner) [inline training_config/mnli_roberta.jsonnet]: I was just checking Anthony's results versus the numbers you're reporting.  You're about 1.5 points below Anthony's MNLI result, and one point below the published number.  I looked at Anthony's code, and it looks like the only difference is that he doesn't have this feedforward layer in there.  What happens if you remove this?

File: training_config/mnli_roberta.jsonnet [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+  "trainer": {
+    "num_epochs": 10,
+    "cuda_device" : 0,
+    "validation_metric": "+accuracy",
+    "learning_rate_scheduler": {
+      "type": "slanted_triangular",
+      "cut_frac": 0.06
+    },
+    "optimizer": {
+      "type": "huggingface_adamw",
```
Comment (matt-gardner) [inline training_config/mnli_roberta.jsonnet]: This is another difference with Anthony's code, though I'm not sure how much difference it makes: https://github.com/anthonywchen/generative-nli/blob/b81d4e3e7f635fcc1e5ca09305707e5df983ec88/configs/roberta_large_config.json#L37-L43

File: training_config/mnli_roberta.jsonnet (hunk 4 above)
Comment (ZhaofengWu) [reply training_config/mnli_roberta.jsonnet]: Ah, thanks for the catch. Will update this (and a new coref config as well).

File: training_config/mnli_roberta.jsonnet (hunk 5 above)
Comment (ZhaofengWu) [reply training_config/mnli_roberta.jsonnet]: RoBERTa actually has this feedforward layer (using the same model as BERT) https://github.com/huggingface/transformers/blob/908fa43b543cf52a3238129624f502240725a6a6/src/transformers/modeling_bert.py#L426-L438

Comment (ZhaofengWu) [reply training_config/mnli_roberta.jsonnet]: And I tried w/ vs. w/o this on SST -- didn't make a difference.

File: training_config/mnli_roberta.jsonnet (hunk 6 above)
Comment (ZhaofengWu) [reply training_config/mnli_roberta.jsonnet]: Weight decay is different but I took the value from the RoBERTa paper (https://arxiv.org/pdf/1907.11692.pdf, last page last table last column). I doubt 1e-5 vs. 2e-5 LR is making 1.5 difference in F1? Anthony also didn't regularize bias which is probably the right thing to do.

Comment (anthonywchen) [reply training_config/mnli_roberta.jsonnet:61]: Yeah in experiments I tried, the differences in different learning rates are basically negligible.

Comment (anthonywchen) [reply training_config/mnli_roberta.jsonnet:61]: One other difference I see is the betas in ADAMW optimizer. I set mine to be [0.9, 0.98], which is recommended from their paper. 

> Similarly, we found setting β2 = 0.98 to improve stability when training with large batch sizes.

Comment (anthonywchen) [reply training_config/mnli_roberta.jsonnet:61]: One other difference is in the beta parameters in the adamw optimizer. I set mine to be [0.9, 0.98], which is recommended on their examples.

https://github.com/pytorch/fairseq/blob/master/examples/roberta/README.glue.md
